x-pack/plugin/sql/sql-action/src/main/java/org/elasticsearch/xpack/sql/action/AbstractSqlQueryRequest.java

```diff
@@ -117,6 +123,86 @@ public AbstractSqlQueryRequest params(List<SqlTypedParamValue> params) {
         this.params = params;
         return this;
     }
+    
+    private static List<SqlTypedParamValue> parseParams(XContentParser p) throws IOException {
+        List<SqlTypedParamValue> result = new ArrayList<>();
+        Token token = p.currentToken();
+        
+        if (token == Token.START_ARRAY) {
+            Object value = null;
+            String type = null;
+            SqlTypedParamValue previousParam = null;
+            
+            while ((token = p.nextToken()) != Token.END_ARRAY) {
+                XContentLocation loc = p.getTokenLocation();
+                
+                if (token == Token.START_OBJECT) {
+                    // we are at the start of a value/type pair... hopefully
+                    SqlTypedParamValue s = SqlTypedParamValue.fromXContent(p);
+                    /*
+                     * Set the xcontentlocation for the first param just in case the first one doesn't meet the parsing rules
+                     * that are checked later in validateParams method and, also, the xcontentlocation of the param that is 
+                     * different from the previous param in list when it comes to its type being explicitly set or inferred.
+                     */
+                    if ((previousParam != null && previousParam.hasExplicitType() == false) || result.isEmpty()) {
+                        s.tokenLocation(loc);
+                    }
+                    result.add(s);
+                    previousParam = s;
```

Comment:
This code executes in both if/else branches, it might be worth considering taking it out.
Another nit: a `switch` with all the top enums (object, string, number, bool, null) might also arguably simplify the code.

Reply:
I've extracted the common code, but I kept the if/else approach. I find it a bit more clear than a `switch`.